binaries/cuprated/src/rpc/json.rs

```diff
@@ -83,212 +125,1126 @@ pub(super) async fn map_request(
+    let c = context.unchecked_blockchain_context();
+
+    let major_version = c.current_hf.as_u8();
+    let height = usize_to_u64(c.chain_height);
+    let prev_id = hex::encode(c.top_hash);
+    let seed_hash = hex::encode(c.top_hash);
+    let difficulty = format!("{:#x}", c.next_difficulty);
+    let median_weight = usize_to_u64(c.median_weight_for_block_reward);
+    let already_generated_coins = c.already_generated_coins;
+    let tx_backlog = txpool::backlog(&mut state.txpool_read)
+        .await?
+        .into_iter()
+        .map(|entry| GetMinerDataTxBacklogEntry {
+            id: hex::encode(entry.id),
+            weight: entry.weight,
+            fee: entry.fee,
+        })
+        .collect();
+
+    Ok(GetMinerDataResponse {
+        base: ResponseBase::OK,
+        major_version,
+        height,
+        prev_id,
+        seed_hash,
+        difficulty,
+        median_weight,
+        already_generated_coins,
+        tx_backlog,
+    })
 }
 
+/// <https://github.com/monero-project/monero/blob/cc73fe71162d564ffda8e549b79a350bca53c454/src/rpc/core_rpc_server.cpp#L3453-L3476>
 async fn prune_blockchain(
-    state: CupratedRpcHandler,
+    mut state: CupratedRpcHandler,
     request: PruneBlockchainRequest,
 ) -> Result<PruneBlockchainResponse, Error> {
-    todo!()
+    let pruned = blockchain_manager::pruned(&mut state.blockchain_manager).await?;
+    let pruning_seed = blockchain_manager::prune(&mut state.blockchain_manager)
+        .await?
+        .compress();
+
+    Ok(PruneBlockchainResponse {
+        base: ResponseBase::OK,
+        pruned,
+        pruning_seed,
+    })
 }
 
+/// <https://github.com/monero-project/monero/blob/cc73fe71162d564ffda8e549b79a350bca53c454/src/rpc/core_rpc_server.cpp#L2035-L2070>
 async fn calc_pow(
-    state: CupratedRpcHandler,
+    mut state: CupratedRpcHandler,
     request: CalcPowRequest,
 ) -> Result<CalcPowResponse, Error> {
-    todo!()
+    let hardfork = HardFork::from_version(request.major_version)?;
+    let block_blob: Vec<u8> = hex::decode(request.block_blob)?;
+    let block = Block::read(&mut block_blob.as_slice())?;
+    let seed_hash = helper::hex_to_hash(request.seed_hash)?;
+
+    // let block_weight = todo!();
+
+    // let median_for_block_reward = blockchain_context::context(&mut state.blockchain_context)
+    //     .await?
+    //     .unchecked_blockchain_context()
+    //     .context_to_verify_block
+    //     .median_weight_for_block_reward;
+
+    // if cuprate_consensus_rules::blocks::check_block_weight(block_weight, median_for_block_reward)
+    //     .is_err()
+    // {
+    //     return Err(anyhow!("Block blob size is too big, rejecting block"));
+    // }
+
+    // TODO: will `CalculatePow` do the above checks?
+    let pow_hash = blockchain_context::calculate_pow(
+        &mut state.blockchain_context,
+        hardfork,
+        block,
+        seed_hash,
+    )
+    .await?;
```

Comment:
Will `BlockChainContextRequest::CalculatePow` do the sanity checks here? If not, what would be the cheapest way to calculate `block_weight` here?